tests/devices/test_default_clifford.py

```diff
@@ -0,0 +1,318 @@
+
+
+@pytest.mark.parametrize("circuit", [circuit_1])
+@pytest.mark.parametrize(
+    "expec_op",
+    [
+        qml.PauliZ(0),
+        qml.PauliZ(0) @ qml.PauliX(1),
+        qml.Hamiltonian([0.42], [qml.PauliZ(0) @ qml.PauliX(1)]),
+    ],
+)
+def test_expectation_clifford(circuit, expec_op):
+    """Test that the execution of default.clifford is possible and agrees with default.qubit"""
+    dev_c = qml.device("default.clifford")
+    dev_d = qml.device("default.qubit")
+
+    def circuit_fn():
+        circuit()
+        return qml.expval(expec_op)
+
+    qn_c = qml.QNode(circuit_fn, dev_c)
+    qn_d = qml.QNode(circuit_fn, dev_d)
+
+    assert np.allclose(qn_c(), qn_d())
+
+
+@pytest.mark.parametrize("circuit", [circuit_1])
+@pytest.mark.parametrize("state", ["tableau", "state_vector"])
+def test_state_clifford(circuit, state):
+    """Test that the execution of default.clifford is possible and agrees with default.qubit"""
+    dev_c = qml.device("default.clifford", state=state)
+    dev_d = qml.device("default.qubit")
+
+    def circuit_fn():
+        circuit()
+        return qml.state()
+
+    qn_c = qml.QNode(circuit_fn, dev_c)
+    qn_d = qml.QNode(circuit_fn, dev_d)
+
+    if state == "state_vector":
+        st1, st2 = qn_c(), qn_d()
+        phase = qml.math.divide(
+            st1, st2, out=qml.math.zeros_like(st1, dtype=complex), where=st1 != 0
+        )[qml.math.nonzero(np.round(st1, 10))]
+        if not qml.math.allclose(phase, 0.0):
+            assert qml.math.allclose(phase / phase[0], qml.math.ones(len(phase)))
+    else:
+        tableau = np.array([[0, 1, 1, 0, 0], [1, 0, 1, 1, 1], [0, 0, 0, 1, 0], [1, 0, 0, 1, 1]])
+        assert qml.math.allclose(tableau, qn_c())
+
+
+@pytest.mark.parametrize("circuit", [circuit_1])
+@pytest.mark.parametrize(
+    "meas_type",
+    ["dm", "purity"],
+)
+def test_meas_clifford(circuit, meas_type):
+    """Test that the execution of default.clifford is possible and agrees with default.qubit"""
+    dev_c = qml.device("default.clifford")
+    dev_d = qml.device("default.qubit")
+
+    def circuit_fn():
+        circuit()
+        return qml.density_matrix([0, 1]) if meas_type == "dm" else qml.purity([0, 1])
+
+    qn_c = qml.QNode(circuit_fn, dev_c)
+    qn_d = qml.QNode(circuit_fn, dev_d)
+
+    assert np.allclose(qn_c(), qn_d())
+
+
+@pytest.mark.parametrize("circuit", [circuit_1])
+def test_prep_snap_clifford(circuit):
+    """Test that the execution of default.clifford is possible and agrees with default.qubit"""
+    dev_c = qml.device("default.clifford", check_clifford=False)
+    dev_d = qml.device("default.qubit")
+
+    def circuit_fn():
+        qml.BasisState(np.array([1, 1]), wires=range(2))
+        circuit()
+        qml.Snapshot(tag="cliffy")
+        return qml.expval(qml.PauliZ(0))
+
+    qn_c = qml.QNode(circuit_fn, dev_c)
+    qn_d = qml.QNode(circuit_fn, dev_d)
+
+    assert np.allclose(qn_c(), qn_d())
+    assert np.allclose(qml.grad(qn_c)(), qml.grad(qn_d)())
```

Comment:
Consider having a separate gradient test.

Reply:
Separated the gradient test now.